codec/unknownproto/unknown_fields.go

```diff
@@ -22,38 +22,45 @@ type descriptorIface interface {
 	Descriptor() ([]byte, []int)
 }
 
-type Checker struct {
-	// AllowUnknownNonCriticals when set will skip over non-critical fields that are unknown.
-	AllowUnknownNonCriticals bool
+// RejectUnknownFieldsStrict rejects any bytes bz with an error that has unknown fields for the provided proto.Message type.
+// This function traverses inside of messages nested via google.protobuf.Any. It does not do any deserialization of the proto.Message.
+func RejectUnknownFieldsStrict(bz []byte, msg proto.Message) error {
+	_, err := RejectUnknownFields(bz, msg, false)
+	return err
 }
 
-func (ckr *Checker) RejectUnknownFields(b []byte, msg proto.Message) error {
-	if len(b) == 0 {
-		return nil
+// RejectUnknownFields rejects any bytes bz with an error that has unknown fields for the provided proto.Message type with an
+// option to allow non-critical fields (specified as those fields with bit 11) to pass through. In either case, the
+// hasUnknownNonCriticals will be set to true if non-critical fields were encountered during traversal. This flag can be
+// used to treat a message with non-critical field different in different security contexts (such as transaction signing).
+// This function traverses inside of messages nested via google.protobuf.Any. It does not do any deserialization of the proto.Message.
+func RejectUnknownFields(bz []byte, msg proto.Message, allowUnknownNonCriticals bool) (hasUnknownNonCriticals bool, err error) {
+	if len(bz) == 0 {
+		return hasUnknownNonCriticals, nil
```

Comment:
For the failing cases, I'd encourage using inlined values for clarity reading, so
```go
if len(bz) == 0 {
		return false, nil
```

Reply:
I originally did it the way you're suggesting, but I changed it to always return `hasUnknownNonCriticals` instead of a constant `false`. The reason is because we always want to return the existing value of `hasUnknownNonCriticals`. While it's obviously false before entering the for loop, as soon as the for loop is entered `hasUnknownNonCriticals` and not `false` needs to be returned. I'd prefer to leave it as is for consistency unless you have a strong objection.

Reply:
Ack, not a problem, understood.